src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -44262,38 +44262,153 @@ static void F((object, object?) t)
             //Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "((x, _) = t).Item2").WithLocation(7, 9));
         }
 
-        [Fact]
+        [Fact, WorkItem(29635, "https://github.com/dotnet/roslyn/issues/29635")]
         public void Discard_02()
         {
-            // https://github.com/dotnet/roslyn/issues/29635 Need to re-infer discards
             var source =
 @"#nullable disable
 class C<T>
 {
 #nullable enable
     void F(object o1, object? o2, C<object> o3, C<object?> o4)
     {
-        _ /*T:object?*/ = o1;
+        _ /*T:object!*/ = o1;
         _ /*T:object?*/ = o2;
-        _ /*T:C<object!>?*/ = o3;
-        _ /*T:C<object?>?*/ = o4;
+        _ /*T:C<object!>!*/ = o3;
+        _ /*T:C<object?>!*/ = o4;
     }
 #nullable disable
     void F(C<object> o)
     {
-        _ /*T:C<object>?*/ = o;
+        _ /*T:C<object>!*/ = o;
     }
 }";
             var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
             comp.VerifyDiagnostics();
             comp.VerifyTypes();
         }
 
+        [Fact, WorkItem(36503, "https://github.com/dotnet/roslyn/issues/36503")]
+        public void Discard_Reinferred()
+        {
+            var source = @"
+#nullable enable
+interface I<in T> {}
+
+class C
+{
+    void M(I<string?> i1, I<string> i2)
+    {
+        var x = i2 ?? i1;
+        _ = x /*T:I<string!>!*/;
+
+        _ /*T:I<string!>!*/ = i2 ?? i1;
+
+        var y = (_ = i2 ?? i1);
+        _ = y /*T:I<string!>!*/;
+    }
+}";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics();
+            comp.VerifyTypes();
+        }
+
+        [Fact, WorkItem(36503, "https://github.com/dotnet/roslyn/issues/36503")]
+        public void Discard_Reinferred_Nested()
+        {
+            var source = @"
+#nullable enable
+interface I<in T> {}
+
+class C
+{
+    void M(I<I<string?>> i1, I<I<string>?> i2)
+    {
+        var x = i2 ?? i1;
```

Comment:
I'm struggling to understand why this works. It seems like the innermost `string` -> `string?` conversion is not safe?

Reply:
I have to say I have trouble with complex cases like this too. The important thing which I wanted to verify here is that `_` works like `var` declaration.